### ISSUE TITLE
GetSrcSetUrls throwing exception if images width is less than WidthStep

### ISSUE THREAD

**Issue by mattbrailsford** [Jeavon/Slimsy#5]
When calling GetSrcSetUrls with a width / height greater than the provided images width / height you get an axception

````
Length cannot be less than zero.
Parameter name: length

at System.String.Substring(Int32 startIndex, Int32 length)
   at Slimsy.Slimsy.GetSrcSetUrls(UrlHelper urlHelper, IPublishedContent publishedContent, Int32 width, Int32 height, Nullable`1 imageCropMode, String outputFormat)
   at System.Web.WebPages.WebPageBase.ExecutePageHierarchy()
   at System.Web.Mvc.WebViewPage.ExecutePageHierarchy()
   at System.Web.WebPages.WebPageBase.ExecutePageHierarchy(WebPageContext pageContext, TextWriter writer, WebPageRenderingBase startPage)
   at System.Web.Mvc.RazorView.RenderView(ViewContext viewContext, TextWriter writer, Object instance)
   at System.Web.Mvc.BuildManagerCompiledView.Render(ViewContext viewContext, TextWriter writer)
   at Umbraco.Core.Profiling.ProfilingView.Render(ViewContext viewContext, TextWriter writer)
   at System.Web.Mvc.ViewResultBase.ExecuteResult(ControllerContext context)
   at System.Web.Mvc.ControllerActionInvoker.InvokeActionResult(ControllerContext controllerContext, ActionResult actionResult)
   at System.Web.Mvc.ControllerActionInvoker.InvokeActionResultFilterRecursive(IList`1 filters, Int32 filterIndex, ResultExecutingContext preContext, ControllerContext controllerContext, ActionResult actionResult)
   at System.Web.Mvc.ControllerActionInvoker.InvokeActionResultFilterRecursive(IList`1 filters, Int32 filterIndex, ResultExecutingContext preContext, ControllerContext controllerContext, ActionResult actionResult)````

From what I can tell it's because you do a substring in the return statement that always assumes the string builder contains something, which if your size is bigger than the max width (ie, your image natural width) then it doesn't and so throws and index error as your requested substring will be a length of -1.

**Comment by mattbrailsford** [comment]
Had a closer look, and updated the title of the issue as it seems to boils down that if the image is smaller than the widthStep then this is was causes the issue.

**Comment by Jeavon** [comment]
Ooh, good catch

**Comment by Jeavon** [comment]
p.s. love the new avatar

**Comment by mattbrailsford** [comment]
I've hard coded a fix in my project, and yes, I really shouldn't be attempting to call this on too small an image. But would be nice if it could just ignore it so it doesn't error. Especially given these images are under user control so we all know a small one will be uploaded at some point :)

And thanks re the avatar :D

**Comment by Jeavon** [comment]
I'm thinking this is the fix:

        private static int MaxWidth(IPublishedContent publishedContent)
        {
            // this should be overridable from an appsetting
            var maxWidth = 2048;

            // if publishedContent is a media item we can see if we can get the source image width & height
            if (publishedContent.ItemType == PublishedItemType.Media)
            {
                var sourceWidth = publishedContent.GetPropertyValue<int>(Constants.Conventions.Media.Width);

                // if source width is less than max width then we should stop at source width
                if (sourceWidth < maxWidth)
                {
                    maxWidth = sourceWidth;
                }

                // if the source image is less than the step then max width should be the first step
                if (maxWidth < WidthStep())
                {
                    maxWidth = WidthStep();
                }
            }

            return maxWidth;
        }

**Comment by mattbrailsford** [comment]
Looks good to me 👍

**Comment by Jeavon** [comment]
Test and committed, thanks!

**Comment by Jeavon** [comment]
It's available in the 2.0.0-beta3-000135 prerelease on MyGet

**Comment by mattbrailsford** [comment]
Sweet! Thanks for the speedy fix. 🚀

**Comment by Jeavon** [comment]
Thanks for running with the alphas/betas for so long! I hope we can move to release before too long

**Comment by mattbrailsford** [comment]
It's been working flawlessly otherwise, so you've done a grand job if that's all I can find :)

**Comment by Jeavon** [comment]
Released in 2.0.0-beta3
